src/Raven.Server/ServerWide/ClusterStateMachine.cs

```diff
@@ -1439,6 +1571,57 @@ private static unsafe (string, BlittableJsonReaderObject) GetCurrentItem(Transac
             return (key, doc);
         }
 
+        public BlittableJsonReaderObject GetCertificateByPrimaryKey(TransactionOperationContext context, string key)
+        {
+            var certs = context.Transaction.InnerTransaction.OpenTable(CertificatesSchema, CertificatesSlice);
+
+            using (Slice.From(context.Allocator, key, out var k))
+            {
+                var tvh = new Table.TableValueHolder();
+                if (certs.ReadByKey(k, out tvh.Reader) == false)
+                    return null;
+                return GetCertificate(context, tvh).Item2;
+            }
+        }
+
+        private static unsafe (string Key, BlittableJsonReaderObject Cert) GetCertificate(TransactionOperationContext context, Table.TableValueHolder result)
+        {
+            var ptr = result.Reader.Read((int)CertificatesTable.Data, out int size);
+            var doc = new BlittableJsonReaderObject(ptr, size, context);
+            var key = Encoding.UTF8.GetString(result.Reader.Read((int)CertificatesTable.Key, out size), size);
+
+            Transaction.DebugDisposeReaderAfterTransaction(context.Transaction.InnerTransaction, doc);
+            return (key, doc);
+        }
+
+        private static unsafe CertificateDefinition GetCertificateDefinition(TransactionOperationContext context, Table.TableValueHolder result)
+        {
+            return JsonDeserializationServer.CertificateDefinition(GetCertificate(context, result).Cert);
+        }
+
+        public List<CertificateDefinition> GetCertificatesByPinningHashSortedByExpiration(TransactionOperationContext context, string hash)
+        {
+
+            var list = GetCertificatesByPinningHash(context, hash).ToList();
+            list.Sort((x,y) => DateTime.Compare(x.NotAfter, y.NotAfter));
+            return list;
+        }
+
+        public IEnumerable<CertificateDefinition> GetCertificatesByPinningHash(TransactionOperationContext context, string hash)
+        {
+            var certs = context.Transaction.InnerTransaction.OpenTable(CertificatesSchema, CertificatesSlice);
+
+            using (Slice.From(context.Allocator, hash, out Slice hashSlice))
+            {
+                foreach (var tvr in certs.SeekForwardFrom(CertificatesSchema.Indexes[CertificatesHashSlice], hashSlice, 0))
+                {
+                    var def = GetCertificateDefinition(context, tvr.Result);
+                    if (def.PublicKeyPinningHash.Equals(hash))
```

Comment:
Inverse this, if this isn't right, break immediately

Reply:
done